src/main/scala/Subtyping.scala

```diff
@@ -0,0 +1,51 @@
+package fuselang
+
+import scala.math.{max,log10,ceil}
+import Syntax._
+import Errors.NoJoin
+
+/**
+ * Subtyping relations are only defined over the number hierarchy. Read 't1 < t2'
+ * as t1 is subtype of t2. The subtyping hierarchy is:
+ *
+ * sized(n) > sized(n - 1) ... > sized(1) > idx > static
+ *
+ * Note that all idx types and static types are equal to each other.
+ */
+object Subtyping {
+  def bitsNeeded(n: Int) = n match {
+    case 0 => 1
+    case n => ceil(log10(n)/log10(2)).toInt + 1
+  }
+
+  def areEqual(t1: Type, t2: Type) = (t1, t2) match {
+    case (_:TStaticInt, _:TStaticInt) => true
```

Comment:
Perhaps this should be:

```scala
case (_:TStaticInt(v1), _:TStaticInt(v2)) => v1 == v2
```

i.e., static int types are equal when the numbers are equal (and unordered otherwise).

Reply:
That'll disallow `1 == 2` since `1 : static(1)` and `2 : static(2)`.

I think I might be conflating subtyping and joins. In this case, we really just want a join on the two static ints. Any pointers on which one should be used where?

Reply:
Good question. One way of looking at is imagining that you have a polymorphic function `equal[T](lhs: T, rhs: T)`. Then you have a call to this function without a type parameter, like `equal(1, 2)`, where the type `T` needs to be inferred. The answer to the question, "is there a type `T` that makes this call well-typed?" is equivalent to "do the types of `1` and `2` have a least upper bound, i.e., a join?"

For this particular case, following this philosophy would say "I can compare 1 and 2 because I can subsume them both to `bit<2>` and do the comparison as I would on `bit<2>` values." This would also nicely handle comparisons between static and non-static numbers.